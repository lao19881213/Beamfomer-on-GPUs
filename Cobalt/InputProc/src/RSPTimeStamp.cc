//# RSPTimeStamp.cc: Small class to hold the timestamps from RSP
//# Copyright (C) 2008-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: RSPTimeStamp.cc 25497 2013-06-27 05:36:14Z mol $

#include <lofar_config.h>

#include "RSPTimeStamp.h"

#include <Common/lofar_iostream.h>
#include <Common/lofar_iomanip.h>

#include <time.h>

namespace LOFAR
{
  namespace Cobalt
  {

    ostream &operator << (ostream &os, const TimeStamp &ts)
    {
      double time_d = ts.getSeconds();
      time_t seconds = static_cast<time_t>(floor(time_d));
      unsigned ms = static_cast<unsigned>(floor((time_d - seconds) * 1000 + 0.5));

      char   buf[26];
      struct tm tm;

      gmtime_r(&seconds, &tm);
      size_t len = strftime(buf, sizeof buf, "%F %T", &tm);
      buf[len] = '\0';

      return os << "[" << ts.getSeqId() << "s, " << ts.getBlockId() << "] = " << buf << "." << setfill('0') << setw(3) << ms << " UTC";
    }

  } // namespace Cobalt
} // namespace LOFAR

