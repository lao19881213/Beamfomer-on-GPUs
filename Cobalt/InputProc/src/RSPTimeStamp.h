//# RSPTimeStamp.h: Small class to hold the timestamps from RSP
//# Copyright (C) 2008-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: RSPTimeStamp.h 26336 2013-09-03 10:01:41Z mol $

#ifndef LOFAR_COBALT_INPUTPROC_RSPTIMESTAMP_H
#define LOFAR_COBALT_INPUTPROC_RSPTIMESTAMP_H

#include <Common/lofar_iosfwd.h>
#include <Common/LofarTypes.h>
#include <Common/LofarLogger.h>

#define EVEN_SECOND_HAS_MORE_SAMPLES

namespace LOFAR
{
  namespace Cobalt
  {

    class TimeStamp
    {
    public:
      TimeStamp(); // empty constructor to be able to create vectors of TimeStamps
      TimeStamp(uint64 time); // for conversion from ints, used to convert values like 0x7FFFFFFF and 0x0 for special cases.
      TimeStamp(uint64 time, unsigned clockSpeed);
      TimeStamp(unsigned seqId, unsigned blockId, unsigned clockSpeed);

      TimeStamp     &setStamp(unsigned seqId, unsigned blockId);
      unsigned      getSeqId() const;
      unsigned      getBlockId() const;
      unsigned      getClock() const
      {
        return itsClockSpeed;
      }

      template <typename T>
      TimeStamp &operator += (T increment);
      template <typename T>
      TimeStamp &operator -= (T decrement);
      TimeStamp operator ++ (int);                        // postfix
      TimeStamp &operator ++ ();                          // prefix
      TimeStamp operator -- (int);
      TimeStamp &operator -- ();

      template <typename T>
      TimeStamp operator +  (T) const;
      template <typename T>
      TimeStamp operator -  (T) const;
      uint64 operator -  (const TimeStamp &) const;

      bool operator >  (const TimeStamp &) const;
      bool operator <  (const TimeStamp &) const;
      bool operator >= (const TimeStamp &) const;
      bool operator <= (const TimeStamp &) const;
      bool operator == (const TimeStamp &) const;
      bool operator != (const TimeStamp &) const;
      bool operator !  () const;

      double getSeconds() const;
      operator uint64 () const;
      operator struct timespec () const;

      friend ostream &operator << (ostream &os, const TimeStamp &ss);

    protected:
      uint64 itsTime;
      unsigned itsClockSpeed;
    };

    inline TimeStamp::TimeStamp() :
      itsTime(0),
      itsClockSpeed(1)
    {
    }

    inline TimeStamp::TimeStamp(uint64 time) :
      itsTime(time),
      itsClockSpeed(1)
    {
    }

    inline TimeStamp::TimeStamp(uint64 time, unsigned clockSpeed) :
      itsTime(time),
      itsClockSpeed(clockSpeed)
    {
      ASSERT(clockSpeed > 0);
    }

    inline TimeStamp::TimeStamp(unsigned seqId, unsigned blockId, unsigned clockSpeed)
    {
      ASSERT(clockSpeed > 0);

      itsClockSpeed = clockSpeed;

#ifdef EVEN_SECOND_HAS_MORE_SAMPLES
      itsTime = ((uint64) seqId * itsClockSpeed + 512) / 1024 + blockId;
#else
      itsTime = ((uint64) seqId * itsClockSpeed) / 1024 + blockId;
#endif
    }

    inline TimeStamp &TimeStamp::setStamp(unsigned seqId, unsigned blockId)
    {
#ifdef EVEN_SECOND_HAS_MORE_SAMPLES
      itsTime = ((uint64) seqId * itsClockSpeed + 512) / 1024 + blockId;
#else
      itsTime = ((uint64) seqId * itsClockSpeed) / 1024 + blockId;
#endif
      return *this;
    }

    inline unsigned TimeStamp::getSeqId() const
    {
#ifdef EVEN_SECOND_HAS_MORE_SAMPLES
      return (unsigned) (1024 * itsTime / itsClockSpeed);
#else
      return (unsigned) ((1024 * itsTime + 512) / itsClockSpeed);
#endif
    }

    inline unsigned TimeStamp::getBlockId() const
    {
#ifdef EVEN_SECOND_HAS_MORE_SAMPLES
      return (unsigned) (1024 * itsTime % itsClockSpeed / 1024);
#else
      return (unsigned) ((1024 * itsTime + 512) % itsClockSpeed / 1024);
#endif
    }

    template <typename T>
    inline TimeStamp &TimeStamp::operator += (T increment)
    {
      itsTime += increment;
      return *this;
    }

    template <typename T>
    inline TimeStamp &TimeStamp::operator -= (T decrement)
    {
      itsTime -= decrement;
      return *this;
    }

    inline TimeStamp &TimeStamp::operator ++ ()
    {
      ++itsTime;
      return *this;
    }

    inline TimeStamp TimeStamp::operator ++ (int)
    {
      TimeStamp tmp = *this;
      ++itsTime;
      return tmp;
    }

    inline TimeStamp &TimeStamp::operator -- ()
    {
      --itsTime;
      return *this;
    }

    inline TimeStamp TimeStamp::operator -- (int)
    {
      TimeStamp tmp = *this;
      --itsTime;
      return tmp;
    }

    template <typename T>
    inline TimeStamp TimeStamp::operator + (T increment) const
    {
      return TimeStamp(itsTime + increment, itsClockSpeed);
    }

    template <typename T>
    inline TimeStamp TimeStamp::operator - (T decrement) const
    {
      return TimeStamp(itsTime - decrement, itsClockSpeed);
    }

    inline uint64 TimeStamp::operator - (const TimeStamp &other) const
    {
      return itsTime - other.itsTime;
    }

    inline bool TimeStamp::operator ! () const
    {
      return itsTime == 0;
    }

    inline double TimeStamp::getSeconds() const
    {
      return 1.0 * itsTime * 1024 / itsClockSpeed;
    }

    inline TimeStamp::operator uint64 () const
    {
      return itsTime;
    }

    inline TimeStamp::operator struct timespec () const
    {
      uint64 ns = (uint64) (getSeconds() * 1e9);
      struct timespec ts;

      ts.tv_sec = ns / 1000000000ULL;
      ts.tv_nsec = ns % 1000000000ULL;

      return ts;
    }

    inline bool TimeStamp::operator > (const TimeStamp &other) const
    {
      return itsTime > other.itsTime;
    }

    inline bool TimeStamp::operator >= (const TimeStamp &other) const
    {
      return itsTime >= other.itsTime;
    }

    inline bool TimeStamp::operator < (const TimeStamp &other) const
    {
      return itsTime < other.itsTime;
    }

    inline bool TimeStamp::operator <= (const TimeStamp &other) const
    {
      return itsTime <= other.itsTime;
    }

    inline bool TimeStamp::operator == (const TimeStamp &other) const
    {
      return itsTime == other.itsTime;
    }

    inline bool TimeStamp::operator != (const TimeStamp &other) const
    {
      return itsTime != other.itsTime;
    }

  } // namespace Cobalt

} // namespace LOFAR

#endif

