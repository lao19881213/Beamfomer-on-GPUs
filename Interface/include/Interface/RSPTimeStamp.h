//#  RSPTimeStamp.h: Small class to hold the timestamps from RSP
//#
//#  Copyright (C) 2006
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: RSPTimeStamp.h 22201 2012-10-03 14:04:42Z mol $

#ifndef LOFAR_INTERFACE_RSPTIMESTAMP_H
#define LOFAR_INTERFACE_RSPTIMESTAMP_H

#include <Common/lofar_iosfwd.h>
#include <Common/LofarTypes.h>
#include <Common/LofarLogger.h>

#define EVEN_SECOND_HAS_MORE_SAMPLES

namespace LOFAR {
  namespace RTCP {

    class TimeStamp {
    public:
      TimeStamp(); // empty constructor to be able to create vectors of TimeStamps
      TimeStamp(int64 time); // for conversion from ints, used to convert values like 0x7FFFFFFF and 0x0 for special cases.
      TimeStamp(int64 time, unsigned clockSpeed);
      TimeStamp(unsigned seqId, unsigned blockId, unsigned clockSpeed);

      TimeStamp	    &setStamp(unsigned seqId, unsigned blockId);
      unsigned	    getSeqId() const;
      unsigned	    getBlockId() const;
      unsigned	    getClock() const { return itsClockSpeed; }

      template <typename T> TimeStamp &operator += (T increment);
      template <typename T> TimeStamp &operator -= (T decrement);
			    TimeStamp  operator ++ (int); // postfix
			    TimeStamp &operator ++ ();	  // prefix
			    TimeStamp  operator -- (int);
			    TimeStamp &operator -- ();

      template <typename T> TimeStamp  operator +  (T) const;
      template <typename T> TimeStamp  operator -  (T) const;
			    int64      operator -  (const TimeStamp &) const;

			    bool       operator >  (const TimeStamp &) const;
			    bool       operator <  (const TimeStamp &) const;
			    bool       operator >= (const TimeStamp &) const;
			    bool       operator <= (const TimeStamp &) const;
			    bool       operator == (const TimeStamp &) const;
			    bool       operator != (const TimeStamp &) const;
                            bool       operator !  () const;

				       operator int64 () const;
				       operator struct timespec () const;

      friend ostream &operator << (ostream &os, const TimeStamp &ss);

    protected:
      int64	      itsTime;
      unsigned        itsClockSpeed;
    };

    inline TimeStamp::TimeStamp():
      itsTime(0),
      itsClockSpeed(0)
      {
      }

    inline TimeStamp::TimeStamp(int64 time):
      itsTime(time),
      itsClockSpeed(0)
      {
      }

    inline TimeStamp::TimeStamp(int64 time, unsigned clockSpeed):
      itsTime(time),
      itsClockSpeed(clockSpeed)
      {
      }

    inline TimeStamp::TimeStamp(unsigned seqId, unsigned blockId, unsigned clockSpeed)
      {
        itsClockSpeed = clockSpeed;

#ifdef EVEN_SECOND_HAS_MORE_SAMPLES
	itsTime = ((int64) seqId * itsClockSpeed + 512) / 1024 + blockId;
#else
	itsTime = ((int64) seqId * itsClockSpeed) / 1024 + blockId;
#endif
      }

    inline TimeStamp &TimeStamp::setStamp(unsigned seqId, unsigned blockId)
      {
#ifdef EVEN_SECOND_HAS_MORE_SAMPLES
	itsTime = ((int64) seqId * itsClockSpeed + 512) / 1024 + blockId;
#else
	itsTime = ((int64) seqId * itsClockSpeed) / 1024 + blockId;
#endif
	return *this;
      }

    inline unsigned TimeStamp::getSeqId() const
      {
#ifdef EVEN_SECOND_HAS_MORE_SAMPLES
	return (unsigned) (1024 * itsTime / itsClockSpeed);
#else
	return (unsigned) ((1024 * itsTime + 512) / itsClockSpeed);
#endif
      }

    inline unsigned TimeStamp::getBlockId() const
      {
#ifdef EVEN_SECOND_HAS_MORE_SAMPLES
	return (unsigned) (1024 * itsTime % itsClockSpeed / 1024);
#else
	return (unsigned) ((1024 * itsTime + 512) % itsClockSpeed / 1024);
#endif
      }

    template <typename T> inline TimeStamp &TimeStamp::operator += (T increment)
      { 
	itsTime += increment;
	return *this;
      }

    template <typename T> inline TimeStamp &TimeStamp::operator -= (T decrement)
      { 
	itsTime -= decrement;
	return *this;
      }

    inline TimeStamp &TimeStamp::operator ++ ()
      { 
	++ itsTime;
	return *this;
      }

    inline TimeStamp TimeStamp::operator ++ (int)
      { 
        TimeStamp tmp = *this;
	++ itsTime;
	return tmp;
      }

    inline TimeStamp &TimeStamp::operator -- ()
      { 
	-- itsTime;
	return *this;
      }

    inline TimeStamp TimeStamp::operator -- (int)
      { 
	TimeStamp tmp = *this;
	-- itsTime;
	return tmp;
      }

    template <typename T> inline TimeStamp TimeStamp::operator + (T increment) const
      { 
	return TimeStamp(itsTime + increment, itsClockSpeed);
      }

    template <typename T> inline TimeStamp TimeStamp::operator - (T decrement) const
      { 
	return TimeStamp(itsTime - decrement, itsClockSpeed);
      }

    inline int64 TimeStamp::operator - (const TimeStamp &other) const
      { 
	return itsTime - other.itsTime;
      }

    inline bool TimeStamp::operator ! () const
      {
	return itsTime == 0;
      }

    inline TimeStamp::operator int64 () const
      {
	return itsTime;
      }

    inline TimeStamp::operator struct timespec () const
      {
	int64		ns = (int64) (itsTime * 1024 * 1e9 / itsClockSpeed);
	struct timespec ts;

	ts.tv_sec  = ns / 1000000000ULL;
	ts.tv_nsec = ns % 1000000000ULL;

	return ts;
      }

    inline bool TimeStamp::operator > (const TimeStamp &other) const
      { 
	return itsTime > other.itsTime;
      }

    inline bool TimeStamp::operator >= (const TimeStamp &other) const
      {
	return itsTime >= other.itsTime;
      }

    inline bool TimeStamp::operator < (const TimeStamp &other) const
      { 
	return itsTime < other.itsTime;
      }

    inline bool TimeStamp::operator <= (const TimeStamp &other) const
      {
	return itsTime <= other.itsTime;
      }

    inline bool TimeStamp::operator == (const TimeStamp &other) const
      { 
	return itsTime == other.itsTime;
      }

    inline bool TimeStamp::operator != (const TimeStamp &other) const
      { 
	return itsTime != other.itsTime;
      }

  } // namespace RTCP

} // namespace LOFAR

#endif

