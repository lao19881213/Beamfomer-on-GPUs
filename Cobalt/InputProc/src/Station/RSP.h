//# RSP.h: RSP data format
//# Copyright (C) 2012-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: RSP.h 26881 2013-10-07 07:45:48Z loose $

#ifndef LOFAR_INPUT_PROC_RSP_H
#define LOFAR_INPUT_PROC_RSP_H

#include <cstddef>
#include <complex>

#include <Common/LofarTypes.h>
#include <InputProc/RSPTimeStamp.h>

namespace LOFAR
{
  namespace Cobalt
  {

    // WARNING: All data is in Little Endian format!
    //
    // Note that C++ bit fields are implementation dependent,
    // so we cannot use them.

    /* A structure fit for the maximum payload size. When reading UDP,
     * just read them straight into this struct, and ::read() will return
     * the size of the packet.
     *
     * When reading packets from file, make sure you read the right number
     * of bytes (see packetSize()).
     */

    struct RSP {
      // ----------------------------------------------------------------------
      // Header and payload, in little endian!
      // ----------------------------------------------------------------------

      struct Header {
        // 2: Beamlet Data CoInterface 5.0
        // 3: Beamlet Data CoInterface 6.0 (8- and 4-bit mode support)
        uint8 version;

        // bit (0=LSB)
        //
        // 4:0    RSP board number
        // 5      (reserved, set to 0)
        // 6      0: payload ok, 1: payload has data errors
        // 7      0: 160 MHz     1: 200 MHz
        uint8 sourceInfo1;

        // bit (0=LSB)
        //
        // 1:0    0: 16-bit      1: 8-bit       2: 4-bit
        // 7:2    (reserved, set to 0)
        uint8 sourceInfo2;

        // identifiers
        uint8 configuration;
        uint16 station;

        // number of beamlets, typically at maximum:
        //   16-bit: 61
        //    8-bit: 122
        //    4-bit: 244
        uint8 nrBeamlets;

        // number of Xr+Xi+Yr+Yi samples per beamlet, typically 16
        uint8 nrBlocks;

        // UNIX timestamp in UTC (= # seconds since 1970-01-01 00:00:00)
        // 0xFFFFFFFF = clock not initialised
        uint32 timestamp;

        // Sample offset within the timestamp.
        //
        // 160 MHz: 160M/1024 = 156250 samples/second.
        //
        // 200 MHz: 200M/1024 = 195212.5 samples/second.
        //                      Even seconds have 195213 samples,
        //                      odd seconds have 195212 samples.
        uint32 blockSequenceNumber;
      } header;

      // Payload, allocated for maximum size.
      union Payload {
        char data[8130];

        // samples are structured as samples[nrBeamlets][nrBlocks],
        // so first all blocks of the first beamlet, then all blocks of the second
        // beamlet, etc.
        //
        // for 4-bit mode:
        //  low octet: real      (2's complement)
        // high octet: imaginary (2's complement)

        struct samples16bit_t { int16 Xr, Xi, Yr, Yi;
        } samples16bit[61 * 16];
        struct samples8bit_t { int8 Xr, Xi, Yr, Yi;
        } samples8bit[122 * 16];
        struct samples4bit_t { int8 X, Y;
        } samples4bit[244 * 16];
      } payload;


      // ----------------------------------------------------------------------
      // Helper functions
      // ----------------------------------------------------------------------

      unsigned rspBoard() const
      {
        return header.sourceInfo1 & 0x1F;
      }

      void rspBoard(unsigned nr)
      {
        header.sourceInfo1 &= ~0x1F;
        header.sourceInfo1 |= (nr & 0x1F);
      }

      bool payloadError() const
      {
        return header.sourceInfo1 & 0x40;
      }

      void payloadError(bool error)
      {
        if (error)
          header.sourceInfo1 |= 0x40;
        else 
          header.sourceInfo1 &= ~0x40;
      }

      unsigned clockMHz() const
      {
        return header.sourceInfo1 & 0x80 ? 200 : 160;
      }

      void clockMHz(unsigned freq)
      {
        switch (freq) {
        default:
        case 200:
          header.sourceInfo1 |= 0x80; 
          break;
        case 160:
          header.sourceInfo1 &= ~0x80;
          break;
        }
      }

      unsigned bitMode() const
      {
        switch (header.sourceInfo2 & 0x3) {
        default:
        case 0x0: return 16;
        case 0x1: return 8;
        case 0x2: return 4;
        }
      }

      void bitMode(unsigned mode)
      {
        header.sourceInfo2 &= ~0x3;
        switch (mode) {
        default:
        case 16:
          header.sourceInfo2 |= 0x0;
          break;
        case 8 :
          header.sourceInfo2 |= 0x1;
          break;
        case 4 :
          header.sourceInfo2 |= 0x2;
          break;
        }
      }

      TimeStamp timeStamp() const
      {
        return TimeStamp(header.timestamp, header.blockSequenceNumber, clockMHz() * 1000000);
      }

      void timeStamp(const TimeStamp& ts)
      {
        header.timestamp = ts.getSeqId();
        header.blockSequenceNumber = ts.getBlockId();
      }

      size_t packetSize() const
      {
        return sizeof(RSP::Header) + header.nrBlocks * header.nrBeamlets * 2 * 2 * bitMode() / 8;
      }


      // ----------------------------------------------------------------------
      // Payload decoding (for debug purposes, assumes data is converted to native
      // endianness)
      // ----------------------------------------------------------------------
      std::complex<int> sample( unsigned beamlet, unsigned block, char polarisation /* 'X' or 'Y' */) const
      {
        const unsigned offset = beamlet * header.nrBlocks + block;

        switch( bitMode() ) {
        default:
        case 16:
          return polarisation == 'X' ? std::complex<int>(payload.samples16bit[offset].Xr,
                                                         payload.samples16bit[offset].Xi)
                 : std::complex<int>(payload.samples16bit[offset].Yr,
                                     payload.samples16bit[offset].Yi);

        case 8:
          return polarisation == 'X' ? std::complex<int>(payload.samples8bit[offset].Xr,
                                                         payload.samples8bit[offset].Xi)
                 : std::complex<int>(payload.samples8bit[offset].Yr,
                                     payload.samples8bit[offset].Yi);

        case 4:
          return polarisation == 'X' ? decode4bit(payload.samples4bit[offset].X)
                 : decode4bit(payload.samples4bit[offset].Y);
        }
      }

    private:

      // decode the 4-bit complex type.
      static std::complex<int> decode4bit( int8 sample )
      {
        // intermediate after << will be int, not int8,
        // so cast to get a signed int8 value.
        int8 re = (int8)(sample << 4) >> 4; // preserve sign
        int8 im =       (sample     ) >> 4; // preserve sign

        // balance range to [-7..7], subject to change!
        if (re == -8) re = -7;
        if (im == -8) im = -7;

        return std::complex<int>(re, im);
      }
    };

  } // namespace Cobalt
} // namespace LOFAR

#endif

