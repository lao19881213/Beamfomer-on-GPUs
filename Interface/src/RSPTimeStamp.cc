//#  RSPTimeStamp.cc: Small class to hold the timestamps from RSP
//#
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: RSPTimeStamp.cc 15255 2010-03-17 13:53:24Z mol $

#include <lofar_config.h>

#include <Interface/RSPTimeStamp.h>
#include <Common/lofar_iostream.h>

namespace LOFAR {
namespace RTCP {

ostream &operator << (ostream &os, const TimeStamp &ts)
{
  return os << "[" << ts.getSeqId() << "s, " << ts.getBlockId() << "]";
}

} // namespace RTCP
} // namespace LOFAR
