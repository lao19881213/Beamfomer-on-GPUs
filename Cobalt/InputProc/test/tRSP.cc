//# tRSP.cc: stand-alone test program for RSP packet structure
//# Copyright (C) 2012-2013  ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id: tRSP.cc 25497 2013-06-27 05:36:14Z mol $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <string>
#include <iostream>

#include <Common/LofarLogger.h>
#include <Common/DataConvert.h>
#include <Stream/FileStream.h>

#include <InputProc/RSPTimeStamp.h>
#include <InputProc/Station/RSP.h>


using namespace LOFAR;
using namespace LOFAR::Cobalt;
using namespace std;

void report( const string &filename )
{
  cout << "---- Checking " << filename << endl;
  FileStream f(filename);

  struct RSP packet;

  // read header
  f.read( &packet.header, sizeof (RSP::Header) );

#ifdef WORDS_BIGENDIAN
  dataConvert(LittleEndian, packet.header.configuration);
  dataConvert(LittleEndian, packet.header.timestamp);
  dataConvert(LittleEndian, packet.header.blockSequenceNumber);
#endif

  cout << "RSP version:  " << (int)packet.header.version << endl;
  cout << "RSP board nr: " << packet.rspBoard() << endl;
  cout << "Payload OK:   " << (packet.payloadError() ? "NO" : "YES") << endl;
  cout << "Clock:        " << packet.clockMHz() << " MHz" << endl;
  cout << "Bit mode:     " << packet.bitMode() << " bit" << endl;
  cout << "Blocks:       " << (int)packet.header.nrBlocks << endl;
  cout << "Beamlets:     " << (int)packet.header.nrBeamlets << endl;

  // read payload
  f.read( &packet.payload, packet.packetSize() - sizeof (RSP::Header) );

#ifdef WORDS_BIGENDIAN
  if (packet.bitMode() == 16)
    dataConvert(LittleEndian, (int16*)&packet.payload, packet.header.nrBlocks * packet.header.nrBeamlets * 2 * 2);
#endif

  cout << "Sample 4 of beamlet 2:  X = " << packet.sample(2, 4, 'X') << endl;
  cout << "Sample 4 of beamlet 2:  Y = " << packet.sample(2, 4, 'Y') << endl;
}

int main()
{
  INIT_LOGGER("tRSP");

  try {
    report( "tRSP.in_16bit" );
    report( "tRSP.in_8bit" );
  } catch (Exception &ex) {
    LOG_FATAL_STR("Caught exception: " << ex);
    return 1;
  }

  return 0;
}

