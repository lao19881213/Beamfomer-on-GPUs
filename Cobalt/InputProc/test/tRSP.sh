#!/bin/sh
./runctest.sh tRSP
