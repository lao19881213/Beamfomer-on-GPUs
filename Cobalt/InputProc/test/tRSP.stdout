---- Checking tRSP.in_16bit
RSP version:  3
RSP board nr: 0
Payload OK:   YES
Clock:        200 MHz
Bit mode:     16 bit
Blocks:       16
Beamlets:     61
Sample 4 of beamlet 2:  X = (-6,-20)
Sample 4 of beamlet 2:  Y = (-18,-41)
---- Checking tRSP.in_8bit
RSP version:  3
RSP board nr: 1
Payload OK:   YES
Clock:        200 MHz
Bit mode:     8 bit
Blocks:       16
Beamlets:     122
Sample 4 of beamlet 2:  X = (3,0)
Sample 4 of beamlet 2:  Y = (0,0)
